src/lib.rs

```diff
@@ -0,0 +1,339 @@
+mod network;
+
+#[allow(dead_code, unused_imports)]
+
+pub type Meters = f32;
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum End {
+    Front,
+    Back,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum Direction {
+    Forward,
+    Backward,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum Directions {
+    Forward,
+    Backward,
+    BothWays,
+    Alternating,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum Side {
+    Left,
+    Right,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum SideDirection {
+    Leftward,
+    Rightward,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum LaneSide {
+    Inside,
+    Outside,
+}
+
+pub enum LaneDirection {
+    Inward,
+    Outward,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum DrivingSide {
+    LHT,
+    RHT,
+}
+
+use Direction::*;
+use DrivingSide::*;
+use End::*;
+use LaneDirection::*;
+use LaneDirection::*;
+use LaneSide::*;
+use Side::*;
+use SideDirection::*;
+
+impl DrivingSide {
+    fn get_direction(&self, side: Side) -> Direction {
+        match (self, side) {
+            (Self::LHT, Side::Left) | (Self::RHT, Side::Right) => Direction::Forward,
+            (Self::LHT, Side::Right) | (Self::RHT, Side::Left) => Direction::Backward,
+        }
+    }
+    fn get_side(&self, dir: Direction) -> Side {
+        match (self, dir) {
+            (Self::LHT, Direction::Forward) | (Self::RHT, Direction::Backward) => Side::Left,
+            (Self::LHT, Direction::Backward) | (Self::RHT, Direction::Forward) => Side::Right,
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum Carriage {
+    Truck,
+    Bus,
+    Car,
+    Bike,
+    Foot,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum Designation {
+    /// Think "carriageway" carriages, anything from trucks to mopeds to drawn carts.
+    Travel(Carriage),
+    /// Loading zones for trucks too, with short stay?
+    Parking(Carriage),
+    /// Verge and stuff, those outdoor eating areas
+    Amenity,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum RoadEdge {
+    /// Not an edge at all, but a continuation into more road surface.
+    /// Expect Join to a Buffer for railings/bollards.
+    Join,
+    /// The road just ends and transitions into another groundcover.
+    Sudden,
+    /// A short rise up from the road surface. Constructed, usually out of concrete in a certain shape.
+    Kerb,
+    /// A cliff or the edge of a bridge.
+    Drop,
+    /// Walls etc. that interrupt the road surface.
+    Barrier,
+}
+
+/// A single lane on the carriageway, with designation, width, etc.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Lane {
+    dir: Directions,
+    designation: Designation,
+    width: Meters,
+    can_enter_from_inside: bool,
+    can_enter_from_outside: bool,
+}
+
+struct CrossWay(Lane);
+
+impl CrossWay {
+    fn unmarked() -> CrossWay {
```

Comment:
Hmm, treating crossings as lanes? I have seen designated foot / cycle crossings now. Is this meant to eventually be part of representing junctions with multi-stage turns or crossing islands and stuff?

Reply:
> Is this meant to eventually be part of representing junctions with multi-stage turns or crossing islands and stuff?

Yes. We start out by thinking about the fully detailed graph of travel ways, which is about as complicated as the OSM graph we will find. Then we try to make sense of the structure of it, bit by bit. Then when we've understood as much as we can, we try to realise what is there (guessing areas, curbs, markings, etc.). Grouping everything into a "Junction" feels like a pretty high level concept, and doesn't have much bearing on stuff like determining routing and interpreting road structure. But a concept that is very cool for stuff like giving verbal directions.

E.g. If we happen to find a footway crossing over a road, that's a pretty obvious crossing even if it's not tagged as such: probably an unmarked crossing, with a gap in the median and lowered curbs, but nothing else.